src/mds/MDSRank.cc

```diff
@@ -2970,16 +2970,22 @@ void MDSRank::command_scrub_start(Formatter *f,
   bool force = false;
   bool recursive = false;
   bool repair = false;
+  bool scrub_mdsdir = false;
   for (auto &op : scrubop_vec) {
     if (op == "force")
       force = true;
     else if (op == "recursive")
       recursive = true;
     else if (op == "repair")
       repair = true;
+    else if (op == "scrub_mdsdir" && path == "/")
+      scrub_mdsdir = true;
   }
 
   std::lock_guard l(mds_lock);
+  if (scrub_mdsdir) {
+    mdcache->enqueue_scrub("~mdsdir", tag, false, true, false, f, on_finish);
+  }
   mdcache->enqueue_scrub(path, tag, force, recursive, repair, f, on_finish);
```

Comment:
As you noted, this could segfault if `on_finish` completion callback is invoked before this is executed. You can use `MDSGather` and create new context for each `enqueue_scrub` invocation and then `activate`.

Reply:
this change has been incorporated. thanks